### ISSUE TITLE
Currently kubekey is not declarative

### ISSUE THREAD

**Issue by siredmar** [kubesphere/kubekey#1999]
### Your current KubeKey version

v3.0.10

### Describe this feature

If my understanding is correct kubekey is not declarative. If kubekey was declarative i would not need to have commands like add node or delete node. I would also be able to easily add or remove nodes by adding or removing them from the Cluster manifest. 

Is there a reason why kubekey isn't implemented this way?

### Describe the solution you'd like

The solution could look something like this:

Start kubekey with the Cluster configuration. Kubekey takes care that the state is created, e.g. `kk ensure -f config.yaml` 
The current state can be read back `kk state -f config.yaml`. 

To be able to make something like this happen, kubekey needs to get the current state from the machines and compares it to the desired state. If something in the Cluster manifest changes, a diff must be created and handeled.

**Comment by pixiake** [comment]
Yes, currently, kk command is used to assist everyone in quickly creating clusters.

KubeKey also has an operator mode, which is declarative and follows the cluster-api specification. You can refer to:
https://github.com/kubesphere/kubekey/blob/master/docs/CAPKK-Developer-Guide.md
https://cluster-api.sigs.k8s.io

**Comment by siredmar** [comment]
@pixiake thanks for giving me the hind to capkk! 
This looks like what i'm searching for. I've looked at the API specs but could not find anything like `put the kubeconfig in this secret if you are done creating the cluster`.  
Am i missing something? How can i reach the cluster after successfully created without logging in manually via SSH and grab that kubeconfig?

**Comment by siredmar** [comment]
I now fully understand what you wrote me. 
However, i have a question. I created a cluster using the resources for Cluster, KubeadmControlPlane, KKCluster, KKMachineTemplate. The k8s cluster gets created fine, but it seems that each KKInstance is provisioned sequentially. 
Is there a way of speeding things up by doing things in parallel? 
Currently my three node control-plane cluster (currently no workers on purpose) takes about 30 minutes to finish with about 10 minutes per node.

I saw the option `kkinstance-concurrency` in main.go. The default is set to 10. So i thought it could handle up to 10 KKinstances in parallel. Am i missing something?
